docs/new-architecture-playbook.md

```diff
@@ -0,0 +1,2182 @@
+---
+id: new-architecture-playbook
+title: The New Architecture Playbook
+---
+
+# Playbook for Adopting the New Architecture
+
+> NOTE: This documentation is still experimental and details are subject to changes as we iterate. Feel free to share your feedback on the [react-native-website PR](https://github.com/facebook/react-native-website) for this page.
+
+This migration guide is designed for React Native **library authors** and **application developers**. It outlines the steps you need to follow to roll out the new Architecture, composed by the **new NativeModule system (TurboModule) and the new Renderer (Fabric)** to your **Android** and **iOS** libraries and apps.
+
+## Table of Contents
+
+The playbook is divided into three sections:
+
+- **Supporting the New Architecture in your Library**
+  - Prerequisites for Supporting the New Architecture in JavaScript
+  - Enabling the New Architecture in your Library
+    - Android, iOS
+- **Supporting the New Architecture in your App**
+  - Prerequisites for Supporting the New Architecture in your App
+  - Enabling the New NativeModule System (TurboModule) in your App
+    - Android, iOS
+  - Enabling the New Renderer (Fabric) in your App
+    - Android, iOS
+- **Appendix**
+
+# Supporting the New Architecture in Your Library
+
+## Prerequisites for Supporting the New Architecture in JavaScript
+
+The following steps will help ensure your modules and components are ready for the new architecture.
+
+### TurboModules: Define Specs in JavaScript
+
+Under the TurboModule system, the JavaScript spec will serve as the source of truth for the methods that are provided by each native module. Without the JavaScript spec, it is up to you to ensure your public method signatures are equivalent on Android and iOS.
+
+As the first step to adopting the new architecture, you will start by creating these specs for your native modules. You can do this, right now, prior to actually migrating your native module library to the new architecture. Your JavaScript spec will be used later on to generate native interface code for all the supported platforms, as a way to enforce uniform APIs across platforms.
+
+#### Writing the JavaScript Spec
+
+The JavaScript spec defines all APIs that are provided by the native module, along with the types of those constants and functions. You will be using Flow to type the input arguments and outputs of your native module’s methods.
+
+By convention, JavaScript spec files are named `Native<MODULE_NAME>.js` and they export a `TurboModuleRegistry` `Spec` object. The following is a basic JavaScript spec template:
```

Comment:
>The following is a basic JavaScript spec template:

This says JavaScript but the syntax below is Flow. It reads to me like an unstated assumption that one would want to use Flow. Flow is very unpopular in open source, though I'm not sure about RN libraries. (Do we have some stats on Flow vs TS there?)

If I understand correctly, there's some TS version of this. Would it be good to separate this right away into "If you use TypeScript..." and "If you use Flow.." sections?

Reply:
Great point! I am not sure if we have data about flow vs. TS in RN libraries, but regardless, we are definitely considering TS support as it's used a lot more widely in open source. Currently, this playbook only support flow so that we can share the progress earlier (as opposed to waiting for the TS support to be ready).

Reply:
To follow up on this specific point: I've added a banner in the playbook that gives some insights on future TS support: 

![Screenshot 2022-01-06 at 16 53 05](https://user-images.githubusercontent.com/3001957/148419971-3ef1e6f7-9e29-473d-a957-0838d7aad26e.png)